add support for DOCKER_CUSTOM_HEADERS env-var (experimental)

thaJeztah opened the pull request:
- related: https://github.com/moby/moby/pull/32150

This environment variable allows for setting additional headers to be sent by the client. Headers set through this environment variable are added to headers set through the config-file (through the HttpHeaders field).

This environment variable can be used in situations where headers must be set for a specific invocation of the CLI, but should not be set by default, and therefore cannot be set in the config-file.

:warning: WARNING: If both config and environment-variable are set, the environment
variable currently overrides all headers set in the configuration file.
This behavior may change in a future update, as we are considering the
environment variable to be appending to existing headers (and to only
override headers with the same name).


<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/docker/cli/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**

**- How I did it**

**- How to verify it**

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog.
It must be placed inside the below triple backticks section:
-->
```markdown changelog
Add support for `DOCKER_CUSTOM_HEADERS` environment variable

This environment variable allows for setting additional headers to be sent by the client. Headers set through this environment variable are added to headers set through the config-file (through the HttpHeaders field).

This environment variable can be used in situations where headers must be set for a specific invocation of the CLI, but should not be set by default, and therefore cannot be set in the config-file.
```

**- A picture of a cute animal (not mandatory but encouraged)**

thaJeztah committed:
add support for DOCKER_CUSTOM_HEADERS env-var (experimental)

This environment variable allows for setting additional headers
to be sent by the client. Headers set through this environment
variable are added to headers set through the config-file (through
the HttpHeaders field).

This environment variable can be used in situations where headers
must be set for a specific invocation of the CLI, but should not
be set by default, and therefore cannot be set in the config-file.

WARNING: If both config and environment-variable are set, the environment
variable currently overrides all headers set in the configuration file.
This behavior may change in a future update, as we are considering the
environment variable to be appending to existing headers (and to only
override headers with the same name).

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

krissetto commented on cli/command/cli_options.go:
```diff
@@ -107,3 +112,93 @@ func WithAPIClient(c client.APIClient) CLIOption {
+		csvReader := csv.NewReader(strings.NewReader(value))
+		fields, err := csvReader.Read()
+		if err != nil {
+			return errdefs.InvalidParameter(errors.Wrapf(err, "failed to set custom headers from %s environment variable", envOverrideHTTPHeaders))
```
From the docs i see these three main error conditions on `csvReader.Read()`:

1) > If the record has an unexpected number of fields, Read returns the record along with the error [ErrFieldCount](https://pkg.go.dev/encoding/csv#ErrFieldCount). 

   Given we'll only have a single "row" of data, `ErrFieldCount` shouldn't really occur;

2) > If the record contains a field that cannot be parsed, Read returns a partial record along with the parse error. The partial record contains all fields read before the error.  

   I don't think we want to manage partial records, so we should return when ending up on that error;


3)  > If there is no data left to be read, Read returns nil, [io.EOF](https://pkg.go.dev/io#EOF).

    We shouldn't ever encounter an empty reader

Given the above, in this error msg shall we mention that the CSV passed in the `DOCKER_CUSTOM_HEADERS` is likely malformed? IIUC that'd be the most likely reason for encountering the error, and might be a bit more helpful to the user than this fairly generic message. WDYT?

thaJeztah replied:
Hm, yes, good point; I was thinking that the CSV error might be useful, but perhaps it's not, and we should just produce a generic "this is likely the wrong format, we expect a comma-separated list of `key=value`.

krissetto replied:
fair point, always try to make user facing errors "explicit" just in case (so they're in our control rather then depending on the library maintainers). No hard feelings if the error it returns is good enough👍

thaJeztah replied:
We should look indeed what the errors look like, and either "replace" all errors it produces, or if it's only a specific type; for that type.

I haven't found the time yet to try what they look like (perhaps we need to add a test-case for it)

thaJeztah replied:
I updated the error to not include the CSV error, and made it more generic

laurazard commented:
I'm wondering about the utility of adding an env var that **completely replaces** headers set via over settings (such as through the config file) vs just appending/overriding headers with the same name. I feel like in most cases, adding headers might be more useful, and if in some cases there's a need to overwrite some values, that can be done on a per-header basis, rather than completely ignoring otherwise configured headers.

krissetto commented on cli/command/cli_options.go:
```diff
@@ -107,3 +112,93 @@ func WithAPIClient(c client.APIClient) CLIOption {
+				return errdefs.InvalidParameter(errors.Errorf("failed to set custom headers from %s environment variable: value contains a key=value pair with an empty key", envOverrideHTTPHeaders))
+			}
+
+			// We don't currently allow key=value pairs, and produce an error.
```
nit: Did you mean to write "... **empty** key=value pairs,  ..."?

thaJeztah replied:
PROBABLY! 🙈

thaJeztah replied:
Fixed :)

krissetto commented on cli/command/cli_options.go:
```diff
@@ -107,3 +112,93 @@ func WithAPIClient(c client.APIClient) CLIOption {
+
+			if v == "" {
+				// Ignore empty values, and consider them to not be set
+				continue
```
I'm not totally sure we should ignore empty values, as it could add some confusion to the user (maybe they were expecting the header to be set but empty). Are there scenarios in which we might expect this to occur?

thaJeztah replied:
My somewhat anticipation here was for `My-Header=${SOME_ENV_VAR}`. I'd be ok with starting with this being an error condition though (which would still keep all options open)

thaJeztah replied:
OK, gave this some thinking, and it's probably fine to just set an empty value, and leave it an exercise to the user to handle `My-Header=${SOME_ENV_VAR}` in situations where `$SOME_ENV_VAR` is not set (they can use the magic of (Ba)sh to handle unset env-vars.

thaJeztah commented:
> I'm wondering about the utility of adding an env var that completely replaces headers set via over settings (such as through the config file) vs just appending/overriding headers with the same name. I feel like in most cases, adding headers might be more useful, and if in some cases there's a need to overwrite some values, that can be done on a per-header basis, rather than completely ignoring otherwise configured headers.

Yes, this is always the tricky one; override vs merge.

I guess in most situations, an env-var would override config (`flag > env-var > config`), but as this is a potential _list_ of options, that tends to make things more of a grey area.

The tricky part with _merging_ is that we would also need a way to _unset_ env-vars; in some parts of our code, I think we have "empty value" meaning "remove", although that one was specifically for User-Agent (allowing removing the user-agent); https://github.com/moby/moby/blob/989426d303a334db4d97f72efe6637d64220d01a/client/request.go#L244-L266

thaJeztah commented:
> > I'm wondering about the utility of adding an env var that completely replaces headers set via over settings (such as through the config file) vs just appending/overriding headers with the same name. I feel like in most cases, adding headers might be more useful, and if in some cases there's a need to overwrite some values, that can be done on a per-header basis, rather than completely ignoring otherwise configured headers.
> 
> Yes, this is always the tricky one; override vs merge.
> 
> I guess in most situations, an env-var would override config (`flag > env-var > config`), but as this is a potential _list_ of options, that tends to make things more of a grey area.
> 
> The tricky part with _merging_ is that we would also need a way to _unset_ env-vars; in some parts of our code, I think we have "empty value" meaning "remove", although that one was specifically for User-Agent (allowing removing the user-agent); https://github.com/moby/moby/blob/989426d303a334db4d97f72efe6637d64220d01a/client/request.go#L244-L266

I updated the PR and added a WARNING to the GoDoc to mention that the behavior may change (headers to be _appended_ in future, and only overwrite same-name ones)

krissetto commented on cli/command/cli_options.go:
```diff
@@ -108,3 +113,107 @@ func WithAPIClient(c client.APIClient) CLIOption {
 	}
 }
+
+// envOverrideHTTPHeaders is the name of the environment variable that can be
```
nit: a bit pedantic, but should we go with only one way of writing these in the docstrings?
- `environment variable`
- `environment-variable`
- `env-var`

thaJeztah replied:
Heh, yeah, I keep making that mistake. Changed to `environment-variable` which seems the most common use, but we should make a pass at looking through our use for consistency

krissetto commented on cli/command/cli_options.go:
```diff
@@ -108,3 +113,107 @@ func WithAPIClient(c client.APIClient) CLIOption {
+// Tuples without a "=" produce an error.
+//
+// It follows CSV rules for escaping, allowing "key=value" pairs to be quoted
+// if they must contain commas. which allows for multiple values for a single
```
nit: we need `,` instead of `.` here

thaJeztah replied:
Ah! fixed

krissetto commented on cli/command/cli_options.go:
```diff
@@ -108,3 +113,107 @@ func WithAPIClient(c client.APIClient) CLIOption {
+			k = strings.TrimSpace(k)
+
+			if k == "" {
+				return errdefs.InvalidParameter(errors.Errorf("failed to set custom headers from %s environment variable: value contains a key=value pair with an empty key", envOverrideHTTPHeaders))
```
could it make sense to print out the wrong key-value in this error message like we do below, to help the user in spotting the bad part of the string? I'm imagining something similar to:  
 
`failed to set custom headers from DOCKER_CUSTOM_HEADERS environment variable: value contains a key=value pair with an empty key (' =somevalue')`  

where the error shown also contains the potential whitespace (so k before trimming the space)

thaJeztah replied:
ah, yes, probably works; let me look

thaJeztah replied:
Added; I decided to change the error-messages to use `'%s'` (not `%q`) to prevent double-escaped quotes in the output

krissetto commented on cli/command/cli_options.go:
```diff
@@ -108,3 +113,107 @@ func WithAPIClient(c client.APIClient) CLIOption {
+
+		if len(env) == 0 {
+			// We should probably not hit this case, as we don't skip values
+			// (only return errors), but if we don't want to discard existing
```
nit: extra `if` this comment?

thaJeztah replied:
fixed!

krissetto reviewed (approved):
overall LGTM, some minor comments

thaJeztah commented on docs/reference/commandline/docker.md:
```diff
@@ -118,13 +118,14 @@ The following list of environment variables are supported by the `docker` comman
 | `DOCKER_CONTENT_TRUST_SERVER` | The URL of the Notary server to use. Defaults to the same URL as the registry.                                                                                                                                                                                    |
 | `DOCKER_CONTENT_TRUST`        | When set Docker uses notary to sign and verify images. Equates to `--disable-content-trust=false` for build, create, pull, push, run.                                                                                                                             |
 | `DOCKER_CONTEXT`              | Name of the `docker context` to use (overrides `DOCKER_HOST` env var and default context set with `docker context use`)                                                                                                                                           |
+| `DOCKER_CUSTOM_HEADERS`       | (Experimental) Configure [custom HTTP headers](#custom-http-headers) to be sent by the client. Headers must be provided as a comma-separated list of `name=value` pairs. This is the equivalent to the `HttpHeaders` field in the configuration file.             |
```
Yeah, I tried to shorten the description here, otherwise the whole table would be re-formatted 😂 

I think we may need to do another pass at docs in a follow-up; I'm keeping this as-is for now, and we can do a follow-up to improve things if that's ok for you.
